packages/block-library/src/image/check-sizes.js

```diff
@@ -0,0 +1,247 @@
+// Utility functions
+function trim( str ) {
+	return str.trim();
+}
+
+function parseNumber( str ) {
+	const num = parseFloat( str );
+	return isNaN( num ) ? null : num;
+}
+
+function parseDescriptor( descriptor ) {
+	const num = parseNumber( descriptor.slice( 0, -1 ) );
+	if ( descriptor.slice( -1 ) === 'w' ) {
+		return { width: num };
+	} else if ( descriptor.slice( -1 ) === 'h' ) {
+		return { height: num };
+	} else if ( descriptor.slice( -1 ) === 'x' ) {
+		return { density: num };
+	}
+	return {};
+}
+
+// Core functions
+function parseSrcset( srcset ) {
+	return srcset
+		.split( ',' )
+		.map( trim )
+		.map( ( source ) => {
+			const [ url, ...descriptors ] = source.split( /\s+/ );
+			const properties = descriptors
+				.map( parseDescriptor )
+				.reduce( ( a, b ) => Object.assign( a, b ), {} );
+			return { url, ...properties };
+		} );
+}
+
+function parseSizes( sizes ) {
+	return sizes
+		.split( ',' )
+		.map( trim )
+		.map( ( size ) => {
+			const [ media, value ] = size.split( /\s+(?=\d)/ );
+			return { media: trim( media ), value: parseNumber( value ) };
+		} );
+}
+
+// WordPress functions
+async function fetchCustomImageSizes( attachmentId ) {
+	try {
+		const response = await fetch(
+			`/wp-json/wp/v2/media/${ attachmentId }`
+		);
+		const data = await response.json();
+
+		if ( data.media_details && data.media_details.sizes ) {
+			const customSizes = Object.entries( data.media_details.sizes ).map(
+				( [ sizeName, sizeData ] ) => {
+					return {
+						name: sizeName,
+						width: sizeData.width,
+						height: sizeData.height,
+					};
+				}
+			);
+
+			return customSizes;
+		}
+	} catch ( error ) {
+		console.error( 'Error fetching custom image sizes:', error );
+	}
+}
+
+// Perform checks and calculations to validate the sizes attribute based on the srcset.
+// Update the `isValid` variable and add any issues to the `reasons` array as necessary.
+async function checkSizes( media ) {
+	const sizes = media?.sizes;
+	const srcset = media?.srcset;
+	const parsedSizes = parseSizes( sizes );
```

Comment:
@lessbloat I'm getting a JavaScript error at this point:

<img width="689" alt="image" src="https://user-images.githubusercontent.com/36432/235213684-b0bb8c22-406e-4a98-8dda-d9106171ca84.png">

I think `media` isn't quite the data you're expecting, though:

<img width="1720" alt="image" src="https://user-images.githubusercontent.com/36432/235213848-0702c372-5e95-40ac-b2db-51ac1f71405b.png">

Any ideas what the issue might be?